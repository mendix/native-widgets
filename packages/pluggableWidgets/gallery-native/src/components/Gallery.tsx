import { createElement, ReactElement, ReactNode } from "react";
import { Text, FlatList, Pressable, View, ViewProps, Platform, TouchableOpacity } from "react-native";
import { ObjectItem, DynamicValue } from "mendix";
import DeviceInfo from "react-native-device-info";
import { GalleryStyle } from "../ui/Styles";
import { PaginationEnum, ScrollDirectionEnum } from "../../typings/GalleryProps";
import { isAvailable } from "@mendix/piw-utils-internal";
import { extractStyles } from "@mendix/pluggable-widgets-tools";

const DEFAULT_RIPPLE_COLOR = "rgba(0, 0, 0, 0.2)";

export interface GalleryProps<T extends ObjectItem> {
    emptyPlaceholder?: ReactNode;
    hasMoreItems: boolean;
    itemRenderer: (renderWrapper: (children: ReactNode, onClick?: () => void) => ReactElement, item: T) => ReactElement;
    items: T[] | undefined;
    loadMoreItems: () => void;
    filters?: ReactNode;
    name: string;
    pagination: PaginationEnum;
    loadMoreButtonCaption?: DynamicValue<string>;
    phoneColumns: number;
    pullDown?: () => void;
    pullDownIsExecuting?: boolean;
    scrollDirection: ScrollDirectionEnum;
    style: GalleryStyle;
    tabletColumns: number;
}

export const Gallery = <T extends ObjectItem>(props: GalleryProps<T>): ReactElement => {
    const isScrollDirectionVertical = props.scrollDirection === "vertical";
    const numColumns = DeviceInfo.isTablet() ? props.tabletColumns : props.phoneColumns;
    const firstItemId = props.items?.[0]?.id;
    const lastItemId = props.items?.[props.items.length - 1]?.id;

    const onEndReached = (): void => {
        if (props.pagination === "virtualScrolling" && props.hasMoreItems) {
            props.loadMoreItems();
        }
    };

    const renderItem = (item: { item: T }): ReactElement =>
        props.itemRenderer((children, onPress) => {
            const listItemWrapperProps: ViewProps = {
                style: isScrollDirectionVertical && { width: `${100 / numColumns}%` },
                testID: `${props.name}-list-item-${item.item.id}`
            };
            const renderListItemContent = (
                <View
                    style={[
                        props.style.listItem,
                        firstItemId === item.item.id && props.style.firstItem,
                        lastItemId === item.item.id && props.style.lastItem
                    ]}
                >
                    {children}
                </View>
            );
            return onPress ? (
                <Pressable {...listItemWrapperProps} onPress={onPress}>
                    {renderListItemContent}
                </Pressable>
            ) : (
                <View {...listItemWrapperProps}>{renderListItemContent}</View>
            );
        }, item.item);

    const loadMoreButton = (): ReactElement | null => {
        const renderButton = (
            <Text style={props.style.loadMoreButtonCaption}>
                {props.loadMoreButtonCaption && isAvailable(props.loadMoreButtonCaption)
                    ? props.loadMoreButtonCaption.value
                    : "Load more"}
            </Text>
        );

        const [pressableRippleProps, loadMoreButtonContainerStyle] = extractStyles(
            props.style.loadMoreButtonPressableContainer,
            ["rippleColor", "borderless", "radius", "foreground"]
        );

        const buttonProps = {
            testID: `${props.name}-pagination-button`,
            onPress: () => props.hasMoreItems && props.loadMoreItems && props.loadMoreItems(),
            style: loadMoreButtonContainerStyle
        };

        return props.pagination === "buttons" && props.hasMoreItems ? (
            Platform.OS === "android" ? (
                <Pressable
                    {...buttonProps}
                    {...(pressableRippleProps
                        ? {
                              android_ripple: {
                                  ...pressableRippleProps,
                                  color: pressableRippleProps.rippleColor ?? DEFAULT_RIPPLE_COLOR
                              }
                          }
                        : {})}
                >
                    {renderButton}
                </Pressable>
            ) : (
                <TouchableOpacity {...buttonProps}>{renderButton}</TouchableOpacity>
            )
        ) : null;
    };

    const renderEmptyPlaceholder = (): ReactElement => (
        <View style={props.style.emptyPlaceholder}>{props.emptyPlaceholder}</View>
    );

    return (
        <View testID={`${props.name}`} style={props.style.container}>
            {props.filters ? <View>{props.filters}</View> : null}
            <FlatList
                {...(isScrollDirectionVertical && props.pullDown ? { onRefresh: props.pullDown } : {})}
                {...(isScrollDirectionVertical ? { numColumns } : {})}
                ListFooterComponent={loadMoreButton}
                ListFooterComponentStyle={{
                    ...props.style.loadMoreButtonContainer,
                    ...(isScrollDirectionVertical ? { marginTop: 8 } : { marginStart: 8 })
                }}
                refreshing={props.pullDownIsExecuting}
                data={props.items}
                horizontal={!isScrollDirectionVertical}
                keyExtractor={item => item.id}
                ListEmptyComponent={renderEmptyPlaceholder}
                onEndReached={onEndReached}
                renderItem={renderItem}
                style={props.style.list}
                testID={`${props.name}-list`}
            />
        </View>
    );
};
