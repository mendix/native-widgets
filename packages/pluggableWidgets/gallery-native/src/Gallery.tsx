import { all } from "deepmerge";
import { createElement, ReactElement, useCallback, useEffect, useMemo, useState, useRef } from "react";
import {
    executeAction,
    FilterType,
    useFilterContext,
    useMultipleFiltering,
    FilterFunction
} from "@mendix/piw-utils-internal";
import { and } from "mendix/filters/builders";
import { defaultGalleryStyle, GalleryStyle } from "./ui/Styles";
import { extractFilters } from "./utils/filters";
import { FilterCondition } from "mendix/filters";
import { Gallery as GalleryComponent, GalleryProps as GalleryComponentProps } from "./components/Gallery";
import { GalleryProps } from "../typings/GalleryProps";
import { ObjectItem } from "mendix";

export const Gallery = (props: GalleryProps<GalleryStyle>): ReactElement => {
    const viewStateFilters = useRef<FilterCondition | undefined>(undefined);
    const [filtered, setFiltered] = useState(false);
    const customFiltersState = useMultipleFiltering();
    const { FilterContext } = useFilterContext();
    const styles = all<GalleryStyle>([defaultGalleryStyle, ...props.style]);
    const currentPage = props.datasource.limit / props.pageSize;

    useEffect(() => {
        if (props.datasource.limit === Number.POSITIVE_INFINITY) {
            props.datasource.setLimit(props.pageSize);
        }
    }, [props.datasource, props.pageSize]);

    useEffect(() => {
        if (props.datasource.filter && !filtered && !viewStateFilters.current) {
            viewStateFilters.current = props.datasource.filter;
        }
    }, [props.datasource, filtered]);

    const filterList = useMemo(
        () => props.filterList.reduce((filters, { filter }) => ({ ...filters, [filter.id]: filter }), {}),
        [props.filterList]
    );

    const initialFilters = useMemo(
        () =>
            props.filterList.reduce(
                (filters, { filter }) => ({
                    ...filters,
                    [filter.id]: extractFilters(filter, viewStateFilters.current)
                }),
                {}
            ),
        [props.filterList, viewStateFilters.current]
    );

    const filters = Object.keys(customFiltersState)
        .map((key: FilterType) => customFiltersState[key][0]?.getFilterCondition())
        .filter((filter): filter is FilterCondition => filter !== undefined);

    if (filters.length > 0) {
        props.datasource.setFilter(filters.length > 1 ? and(...filters) : filters[0]);
    } else if (filtered) {
        props.datasource.setFilter(undefined);
    } else {
        props.datasource.setFilter(viewStateFilters.current);
    }

    const loadMoreItems = useCallback(() => {
        props.datasource.setLimit((currentPage + 1) * props.pageSize);
    }, [currentPage, props.datasource, props.pageSize]);

    const itemRenderer: GalleryComponentProps<ObjectItem>["itemRenderer"] = useCallback(
        (renderWrapper, item) =>
            renderWrapper(
                props.content?.get(item),
                props.onClick ? () => executeAction(props.onClick?.get(item)) : undefined
            ),
        [props.content, props.onClick]
    );

    const pullDown = useCallback(() => props.pullDown && executeAction(props.pullDown), [props.pullDown]);

    const isFilterable = props.filterList.length > 0;
    const filterContextProvider = useMemo(
        () =>
            isFilterable ? (
                <FilterContext.Provider
                    value={{
                        filterDispatcher: (prev: FilterFunction) => {
                            if (prev.filterType) {
                                const [, filterDispatcher] = customFiltersState[prev.filterType];
                                filterDispatcher(prev);
                                setFiltered(true);
                            }
                            return prev;
                        },
                        multipleAttributes: filterList,
                        multipleInitialFilters: initialFilters
                    }}
                >
                    {props.filtersPlaceholder}
                </FilterContext.Provider>
            ) : null,
        [FilterContext, customFiltersState, filterList, initialFilters, isFilterable, props.filtersPlaceholder]
    );

    return (
        <GalleryComponent
            emptyPlaceholder={props.emptyPlaceholder}
            hasMoreItems={props.datasource.hasMoreItems ?? false}
            itemRenderer={itemRenderer}
            items={props.datasource.items ?? []}
            filters={filterContextProvider}
            loadMoreItems={loadMoreItems}
            name={props.name}
            pagination={props.pagination}
            loadMoreButtonCaption={props.loadMoreButtonCaption}
            phoneColumns={props.phoneColumns}
            pullDown={props.pullDown && pullDown}
            pullDownIsExecuting={props.pullDown?.isExecuting ?? false}
            scrollDirection={props.scrollDirection}
            style={styles}
            tabletColumns={props.tabletColumns}
        />
    );
};
