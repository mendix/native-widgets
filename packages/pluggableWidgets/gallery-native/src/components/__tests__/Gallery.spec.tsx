/**
 * @jest-environment jsdom
 */
import { createElement } from "react";
import { Text, TextInput } from "react-native";
import { ObjectItem, GUID, DynamicValue, ValueStatus } from "mendix";
import { render, fireEvent, act } from "@testing-library/react-native";
import { Gallery, GalleryProps } from "../Gallery";

jest.mock("react-native-device-info", () => ({ isTablet: jest.fn().mockReturnValue(false) }));

const itemWrapperFunction =
    ({ onClick }: { onClick?: () => void }): GalleryProps<ObjectItem>["itemRenderer"] =>
    (wrapper, item) =>
        wrapper(item.id, onClick);

const defaultProps: GalleryProps<ObjectItem> = {
    accessible: true,
    hasMoreItems: true,
    itemRenderer: itemWrapperFunction({}),
    items: [{ id: "11" as GUID }, { id: "22" as GUID }, { id: "33" as GUID }],
    loadMoreItems: jest.fn(),
    name: "gallery-test",
    pagination: "virtualScrolling",
    phoneColumns: 2,
    scrollDirection: "vertical",
    style: { container: {} },
    tabletColumns: 3
};

describe("Gallery", () => {
    describe("rendering", () => {
        it("renders correctly", () => {
            const gallery = render(<Gallery {...defaultProps} />);
            expect(gallery).toMatchSnapshot();
        });

        it("renders correctly horizontal", () => {
            const gallery = render(<Gallery {...defaultProps} scrollDirection="horizontal" />);
            expect(gallery).toMatchSnapshot();
        });

        it("renders correctly with empty list and custom placeholder", () => {
            const gallery = render(
                <Gallery
                    {...defaultProps}
                    pullDownIsExecuting={false}
                    emptyPlaceholder={<Text>Empty list...</Text>}
                    hasMoreItems={false}
                    items={[]}
                />
            );
            expect(gallery).toMatchSnapshot();
            const emptyPlaceholder = gallery.getByText("Empty list...");
            expect(emptyPlaceholder).toBeDefined();
        });

        it("renders correctly with filter", () => {
            const gallery = render(<Gallery {...defaultProps} filters={<TextInput />} />);
            expect(gallery).toMatchSnapshot();
        });

        describe("rendering with load more button", () => {
            it("renders correctly", () => {
                const gallery = render(<Gallery {...defaultProps} pagination="buttons" />);
                expect(gallery).toMatchSnapshot();
            });

            it("renders correctly with custom paging button title", () => {
                const loadMoreButtonCaption: DynamicValue<string> = {
                    status: ValueStatus.Available,
                    value: "Show more"
                };
                const gallery = render(
                    <Gallery {...defaultProps} loadMoreButtonCaption={loadMoreButtonCaption} pagination="buttons" />
                );
                expect(gallery).toMatchSnapshot();

                const customPagingText = gallery.getByText("Show more");
                expect(customPagingText).toBeDefined();
            });

            it("it shouldn't render the paging button if hasn't more item", () => {
                const loadMoreButtonCaption: DynamicValue<string> = {
                    status: ValueStatus.Available,
                    value: "Load more"
                };
                const gallery = render(
                    <Gallery
                        {...defaultProps}
                        loadMoreButtonCaption={loadMoreButtonCaption}
                        hasMoreItems={false}
                        pagination="buttons"
                    />
                );
                expect(gallery).toMatchSnapshot();

                const loadMoreButtonText = gallery.queryByText("Load more");
                expect(loadMoreButtonText).toBeNull();
            });
        });
    });

    describe("events", () => {
        it("item on click action", () => {
            const onClick = jest.fn();
            const gallery = render(<Gallery {...defaultProps} itemRenderer={itemWrapperFunction({ onClick })} />);
            const galleryFirstItem = gallery.getByTestId("gallery-test-list-item-22");
            fireEvent.press(galleryFirstItem);
            expect(onClick).toBeCalledTimes(1);
        });

        it("triggers pull down action", async () => {
            const pullDown = jest.fn();
            const galleryComponent = render(
                <Gallery {...defaultProps} scrollDirection="vertical" pullDown={pullDown} pullDownIsExecuting />
            );
            const galleryFlatList = galleryComponent.getByTestId("gallery-test-list");
            expect(galleryFlatList).toBeDefined();

            const { refreshControl } = galleryFlatList.props;
            await act(async () => {
                refreshControl.props.onRefresh();
            });
            expect(pullDown).toBeCalledTimes(1);
        });

        it("triggers load more items events on end reached", () => {
            const gallery = render(<Gallery {...defaultProps} />);
            const galleryList = gallery.getByTestId("gallery-test-list");
            fireEvent.scroll(galleryList, {
                nativeEvent: {
                    contentOffset: {
                        y: 500
                    },
                    contentSize: {
                        height: 500,
                        width: 100
                    },
                    layoutMeasurement: {
                        height: 100,
                        width: 100
                    }
                }
            });
            expect(defaultProps.loadMoreItems).toBeCalledTimes(1);
        });

        it("it shouldn't triggers the load more items event when item list empty", () => {
            const gallery = render(<Gallery {...defaultProps} items={[]} hasMoreItems={false} />);
            const galleryList = gallery.getByTestId("gallery-test-list");
            fireEvent.scroll(galleryList, {
                nativeEvent: {
                    contentOffset: {
                        y: 500
                    },
                    contentSize: {
                        height: 500,
                        width: 100
                    },
                    layoutMeasurement: {
                        height: 100,
                        width: 100
                    }
                }
            });
            expect(defaultProps.loadMoreItems).not.toBeCalled();
        });

        describe("with pagination button", () => {
            it("triggers load more items event on click load more items button", () => {
                const gallery = render(<Gallery {...defaultProps} pagination="buttons" />);
                const loadMoreItemsButton = gallery.getByTestId("gallery-test-pagination-button");
                fireEvent.press(loadMoreItemsButton);
                expect(defaultProps.loadMoreItems).toBeCalledTimes(1);
            });
        });
    });
});
