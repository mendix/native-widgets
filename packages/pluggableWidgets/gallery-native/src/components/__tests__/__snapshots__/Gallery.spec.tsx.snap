// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery rendering rendering with load more button it shouldn't render the paging button if hasn't more item 1`] = `
<View
  style={{}}
  testID="gallery-test"
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    ListFooterComponentStyle={
      {
        "marginTop": 8,
      }
    }
    data={
      [
        {
          "id": "11",
        },
        {
          "id": "22",
        },
        {
          "id": "33",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="gallery-test-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-11"
          >
            <View
              style={
                [
                  undefined,
                  undefined,
                  false,
                ]
              }
            >
              11
            </View>
          </View>
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-22"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  false,
                ]
              }
            >
              22
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-33"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  undefined,
                ]
              }
            >
              33
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "marginTop": 8,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery rendering rendering with load more button renders correctly 1`] = `
<View
  style={{}}
  testID="gallery-test"
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    ListFooterComponentStyle={
      {
        "marginTop": 8,
      }
    }
    data={
      [
        {
          "id": "11",
        },
        {
          "id": "22",
        },
        {
          "id": "33",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="gallery-test-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-11"
          >
            <View
              style={
                [
                  undefined,
                  undefined,
                  false,
                ]
              }
            >
              11
            </View>
          </View>
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-22"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  false,
                ]
              }
            >
              22
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-33"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  undefined,
                ]
              }
            >
              33
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "marginTop": 8,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={{}}
          testID="gallery-test-pagination-button"
        >
          <Text>
            Load more
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery rendering rendering with load more button renders correctly with custom paging button title 1`] = `
<View
  style={{}}
  testID="gallery-test"
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    ListFooterComponentStyle={
      {
        "marginTop": 8,
      }
    }
    data={
      [
        {
          "id": "11",
        },
        {
          "id": "22",
        },
        {
          "id": "33",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="gallery-test-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-11"
          >
            <View
              style={
                [
                  undefined,
                  undefined,
                  false,
                ]
              }
            >
              11
            </View>
          </View>
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-22"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  false,
                ]
              }
            >
              22
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-33"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  undefined,
                ]
              }
            >
              33
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "marginTop": 8,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={{}}
          testID="gallery-test-pagination-button"
        >
          <Text>
            Show more
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery rendering renders correctly 1`] = `
<View
  style={{}}
  testID="gallery-test"
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    ListFooterComponentStyle={
      {
        "marginTop": 8,
      }
    }
    data={
      [
        {
          "id": "11",
        },
        {
          "id": "22",
        },
        {
          "id": "33",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="gallery-test-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-11"
          >
            <View
              style={
                [
                  undefined,
                  undefined,
                  false,
                ]
              }
            >
              11
            </View>
          </View>
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-22"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  false,
                ]
              }
            >
              22
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-33"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  undefined,
                ]
              }
            >
              33
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "marginTop": 8,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery rendering renders correctly horizontal 1`] = `
<View
  style={{}}
  testID="gallery-test"
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    ListFooterComponentStyle={
      {
        "marginStart": 8,
      }
    }
    data={
      [
        {
          "id": "11",
        },
        {
          "id": "22",
        },
        {
          "id": "33",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="gallery-test-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={false}
          testID="gallery-test-list-item-11"
        >
          <View
            style={
              [
                undefined,
                undefined,
                false,
              ]
            }
          >
            11
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={false}
          testID="gallery-test-list-item-22"
        >
          <View
            style={
              [
                undefined,
                false,
                false,
              ]
            }
          >
            22
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={false}
          testID="gallery-test-list-item-33"
        >
          <View
            style={
              [
                undefined,
                false,
                undefined,
              ]
            }
          >
            33
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "marginStart": 8,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery rendering renders correctly with empty list and custom placeholder 1`] = `
<View
  style={{}}
  testID="gallery-test"
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    ListFooterComponentStyle={
      {
        "marginTop": 8,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshing={false}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="gallery-test-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View>
        <Text>
          Empty list...
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "marginTop": 8,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery rendering renders correctly with filter 1`] = `
<View
  style={{}}
  testID="gallery-test"
>
  <View>
    <TextInput />
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    ListFooterComponentStyle={
      {
        "marginTop": 8,
      }
    }
    data={
      [
        {
          "id": "11",
        },
        {
          "id": "22",
        },
        {
          "id": "33",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="gallery-test-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-11"
          >
            <View
              style={
                [
                  undefined,
                  undefined,
                  false,
                ]
              }
            >
              11
            </View>
          </View>
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-22"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  false,
                ]
              }
            >
              22
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "width": "50%",
              }
            }
            testID="gallery-test-list-item-33"
          >
            <View
              style={
                [
                  undefined,
                  false,
                  undefined,
                ]
              }
            >
              33
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "marginTop": 8,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
